### PR TITLE
Add more logging for `gossiper::lock_endpoint` and `storage_service::handle_state_normal`

### PR DESCRIPTION
**Pull request by kbr-scylla**
In a longevity test reported in scylladb/scylladb#16668 we observed that
NORMAL state is not being properly handled for a node that replaced
another node. Either handle_state_normal is not being called, or it is
but getting stuck in the middle. Which is the case couldn't be
determined from the logs, and attempts at creating a local reproducer
failed.

Thus the plan is to continue debugging using the longevity test, but we need
more logs. To check whether `handle_state_normal` was called and which branches
were taken, include some INFO level logs there. Also, detect deadlocks inside
`gossiper::lock_endpoint` by reporting an error message if `lock_endpoint`
waits for the lock for too long.

Ref: scylladb/scylladb#16668

### CODE REVIEW

**Comment by kbr-scylla** [comment]
@bhalevy please review -- I believe this is necessary to proress on debugging https://github.com/scylladb/scylladb/issues/16668 (I don't have any other ideas)

File: service/storage_service.cc [hunk 1]
```diff
@@ -3827,7 +3830,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
         }
         auto current_owner = current->second;
         if (endpoint == current_owner) {
-            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, current_owner, t);
+            slogger.info("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, current_owner, t);
```
Comment (bhalevy) [inline service/storage_service.cc:3833]: per-token printouts, and particularly for the unchanged case are a bit excessive.
why do we need them?

File: gms/gossiper.cc [hunk 2]
```diff
@@ -775,14 +775,14 @@ future<> gossiper::do_status_check() {
     }
 }
 
-gossiper::endpoint_permit::endpoint_permit(endpoint_locks_map::entry_ptr&& ptr, inet_address addr, std::string caller) noexcept
+gossiper::endpoint_permit::endpoint_permit(endpoint_locks_map::entry_ptr&& ptr, inet_address addr, seastar::compat::source_location caller) noexcept
```
Comment (bhalevy) [inline gms/gossiper.cc:778]: ack

File: service/storage_service.cc (hunk 1 above)
Comment (kbr-scylla) [reply service/storage_service.cc:3833]: I want to be able to distinguish the 4 branches.
Currently 2 branches contain an INFO log (which is already per-token) and 2 branches don't, so we don't have a way to distinguish which of the latter was taken.
By adding an INFO log to one of the two branches we're now distinguishing them all.

This branch is fortunately rare. The first branch is the most common one so it makes sense not to add a log message there.

File: gms/gossiper.cc [hunk 3]
```diff
@@ -833,7 +838,38 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep, permi
...
+                    return fmt_loc(*l);
+                };
+                // The abort had to come from `abort_on_expiry`'s timer.
+                logger.error(
+                    "{}: waiting for endpoint lock (ep={}) took more than {}, signifying possible deadlock;"
+                    " holders: {}, first holder: {}, last holder (might not be current): {}",
+                    fmt_loc(l), ep, duration, eptr->holders, fmt_loc_opt(eptr->first_holder), fmt_loc_opt(eptr->last_holder));
+            } else {
+                throw;
+            }
```
Comment (bhalevy) [inline gms/gossiper.cc:870]: nit: I'd reverse the conditions to throw `if (_abort_source.abort_requested())` to reduce nesting.

File: service/storage_service.cc (hunk 1 above)
Comment (bhalevy) [reply service/storage_service.cc:3833]: if it's rare then ok

**Comment by kbr-scylla** [comment]
v2: reverse conditions in `lock_endpoint` to reduce nesting

**Review by bhalevy — APPROVED** [review]

**Pull request merged**